proof-of-spacetime.md

```diff
@@ -0,0 +1,198 @@
+## Proof-of-Spacetime
+
+This document describes Rational-PoSt, the Proof-of-Spacetime used in Filecoin.
+
+## Rational PoSt
+
+### Definitions
+
+| Name | Value |Description |
+|------|-------|------------|
+| `POST_PROVING_PERIOD` | `2880` blocks  (~24h) | The time interval in which a PoSt has to be submitted. |
+| `POST_CHALLENGE_TIME` | `240` blocks (~2h) | The time offset at which the actual work of generating the PoSt should be started. This is some delta before the end of the `Proving Period`, and as such less then a single `Proving Period`. |
+
+{{% notice todo %}}
+**TODO**: The above values are tentative and need both backing from research as well as detailed reasoning why we picked them.
+{{% /notice %}}
+
+### High Level API
+
+#### Fault Detection
+
+Fault detection happens over the course of the life time of a sector. When the sector is for some reason unavailable, the miner is responsible to submit the known `faults` using the `AddFaults` message to the chain.
```

Comment:
Reading this PR, aren't all faults going to be ones that go-filecoin doesn't know about?  Since faults are reset at the beginning of a proving period?  Or is that we could be challenged on the same (possibly faulted) sector more than once in a proving period...

Reply:
"the miner is responsible to submit the known faults..."  *before the PoSt challenge window begins*. If a miner experiences a fault after the PoSt challenge window opens but before computing the PoSt, they will be unable to submit a PoSt in this period, will lose power, and potentially face other penalties.

Reply:
no, the miner just has to submit the faults before they submit their post.

Reply:
I was led to believe that the seed for challenges would be sampled at a fixed chain height, and that faults reported before this height would affect the seed/challenges, but those reported afterwards would not be taken into account (and hence might be challenged). I would love to be wrong, but let's be very clear here.

If I'm wrong, a miner could sample for challenges, discover a challenge they don't like, report a fault, sample again for a different value, 🔁 . Is that ok?

Reply:
Validation expects the challenge to be taken at a specific block height, including the faults. (I need to update somethings in the PR to make this clearer) So you can _not_ change your challenge in any way once you know about it.

Reply:
Note 1: faults can only be submitted before POST_CHALLENGE_TIME. So miners can't wait for the challenge to appear in order to report faults, and as dig is saying the can't affect the challenge once it appears.
Note 2: one does not know the challenge ahead of time, so they can't really try to minimize the reported faults. So, yes, reporting faults would affect the challenge, but they can't predict how, hence not a problem.

Does this clarify?

Reply:
@nicola that doesnt work. What happens if a miner experiences a fault after the PoSt deadline? They won't be able to complete their PoSt or submit a new set of faults, and will just miss their PoSt entirely (bad).

Reply:
We already went through this. The chance that this happens is very small. You should have all of your fault submitted before you see the challenge. If you fault just after you see the challenge and just before you manage to grab the challenged data, then you fail the current post. You get penalized, it’s the equivalent of faulting every sector and then reintroduce them in the next pp.

If we are not fine with this, then PoSt, must get many more challenges, linear in the number of sectors - so quite a lot.

Reply:
@nicola thats not an okay solution

what happens if network delays cause the fault submission to get pushed past the deadline?
what if a miner just doesnt notice their data is corrupted until they try to read it (really common)?
"sorry, you lose everything, piss off" is not a good answer

Reply:
You “lose everything” in this case means: you temporarily loose all power and you get a penalization. This does not trigger storage collateral (would happen if this repeats five times) or pledge collateral loss.

Temporary power reduction seem fine to me

Reply:
Another note: the SLA that we are giving to the users is that the miners will keep on storing the data, if a miner realised during challenge time that they don’t have the data, then they have been either very unlucky or offered bad service, hence this penalisation ain’t that bad

Reply:
but miners get penalized for the faults they report. I havent gotten a good answer for why this needs to be complicated like this. Let miners submit faults for their post up until they submit the post.